prov/rxd/src/rxd_cq.c

```diff
@@ -229,391 +239,523 @@ static void rxd_ep_recv_data(struct rxd_ep *ep, struct rxd_x_entry *rx_entry,
 {
 	struct rxd_x_entry *rx_entry;
-	struct rxd_ctrl_pkt *pkt = rxd_get_ctrl_pkt(rts_pkt);
-	struct dlist_entry *item;
 
 	//TODO - improve this search
-	dlist_foreach(&ep->active_rx_list, item) {
-		rx_entry = container_of(item, struct rxd_x_entry, entry);
+	dlist_foreach_container(&ep->peers[pkt->pkt_hdr.peer].rx_list,
+				struct rxd_x_entry, rx_entry, entry) {
 		if (rx_entry->tx_id == pkt->pkt_hdr.tx_id &&
-		    rx_entry->key == pkt->pkt_hdr.key &&
-		    rx_entry->peer == rts_pkt->peer) {
-			rxd_post_cts(ep, rx_entry, rts_pkt);
-			return 0;
-		}
+		    rx_entry->msg_id == pkt->pkt_hdr.msg_id &&
+		    rx_entry->peer == pkt->pkt_hdr.peer)
+			return rx_entry;
 	}
 
-	return -FI_ENOMSG;
+	return NULL;
 }
 
-static int rxd_check_post_unexp(struct dlist_entry *list,
-				struct rxd_pkt_entry *pkt_entry)
+static void rxd_check_post_unexp(struct rxd_ep *ep, struct dlist_entry *list,
+				 struct rxd_pkt_entry *pkt_entry)
 {
 	struct rxd_pkt_entry *unexp;
-	struct rxd_ctrl_pkt *pkt = rxd_get_ctrl_pkt(pkt_entry);
-	struct rxd_ctrl_pkt *unexp_pkt;
-	struct dlist_entry *item;
-
-	dlist_foreach(list, item) {
-		unexp = container_of(item, struct rxd_pkt_entry, d_entry);
-		unexp_pkt = rxd_get_ctrl_pkt(unexp);
-		if (unexp_pkt->pkt_hdr.tx_id == pkt->pkt_hdr.tx_id &&
-		    unexp_pkt->pkt_hdr.key == pkt->pkt_hdr.key &&
-		    unexp->peer == pkt_entry->peer)
-			return 0;
-	}
-	dlist_insert_tail(&pkt_entry->d_entry, list);
-	return -FI_ENOMSG;
-}
+	struct rxd_op_pkt *unexp_pkt;
+	struct rxd_op_pkt *pkt = (struct rxd_op_pkt *) (pkt_entry->pkt);
 
-void rxd_progress_inline(struct rxd_ep *ep, struct rxd_pkt_entry *pkt_entry,
-			 struct rxd_x_entry *rx_entry)
-{
-	struct fi_cq_err_entry err_entry;
-	struct rxd_cq *rx_cq = rxd_ep_rx_cq(ep);
-	struct util_cntr *cntr = ep->util_ep.rx_cntr;
-	struct rxd_ctrl_pkt *pkt = rxd_get_ctrl_pkt(pkt_entry);
-	uint64_t recvd;
+	if (rxd_env.ooo_rdm)
+		goto insert;
 
-	recvd = ofi_copy_to_iov(rx_entry->iov, rx_entry->iov_count, 0, 
-				pkt->data, pkt->ctrl_hdr.size);
-	rx_entry->next_seg_no++;	
-	rx_entry->key = pkt->pkt_hdr.key;
-	rx_entry->tx_id = pkt->pkt_hdr.tx_id;
-	rx_entry->peer = pkt_entry->peer;
-	rxd_ep_post_ack(ep, rx_entry);
-
-	/* Handle CQ comp */
-	if (recvd == pkt->ctrl_hdr.size) {
-		rx_cq->write_fn(rx_cq, &rx_entry->cq_entry);
-		/* Handle cntr */
-		if (cntr)
-			cntr->cntr_fid.ops->add(&cntr->cntr_fid, 1);
-	} else {
-		memset(&err_entry, 0, sizeof(err_entry));
-		err_entry.op_context = rx_entry->cq_entry.op_context;
-		err_entry.flags = (FI_MSG | FI_RECV);
-		err_entry.len = recvd;
-		err_entry.err = FI_ETRUNC;
-		err_entry.prov_errno = 0;
-		rxd_cq_report_error(rx_cq, &err_entry);
+	dlist_foreach_container(list, struct rxd_pkt_entry, unexp, d_entry) {
+		unexp_pkt = (struct rxd_op_pkt *) (unexp->pkt);
+		if (unexp_pkt->pkt_hdr.tx_id == pkt->pkt_hdr.tx_id &&
+		    unexp_pkt->pkt_hdr.msg_id == pkt->pkt_hdr.msg_id &&
+		    unexp_pkt->pkt_hdr.peer == pkt->pkt_hdr.peer) {
```

Comment:
This series of checks appears a few times, and I think is within a function call.  Can this be isolated to a static inline or macro?